spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/health/PubSubHealthIndicator.java

```diff
@@ -16,55 +16,73 @@
 
-import java.util.UUID;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
 
 import com.google.api.gax.rpc.ApiException;
 import com.google.api.gax.rpc.StatusCode;
 import com.google.api.gax.rpc.StatusCode.Code;
-import com.google.cloud.spring.pubsub.core.PubSubTemplate;
+import com.google.cloud.spring.pubsub.support.AcknowledgeablePubsubMessage;
 
 import org.springframework.boot.actuate.health.AbstractHealthIndicator;
 import org.springframework.boot.actuate.health.Health;
 import org.springframework.util.Assert;
+import org.springframework.util.concurrent.ListenableFuture;
 
 /**
- * Default implemenation of
+ * Default implementation of
  * {@link org.springframework.boot.actuate.health.HealthIndicator} for Pub/Sub. Validates
- * if connection is successful by looking for a random generated subscription name that
- * won't be found. If no subscription is found we know the client is able to connect to
- * GCP Pub/Sub APIs.
+ * if connection is successful by pulling message asynchronously from the pubSubHealthTemplate.
+ * If no subscription is found we know the client is able to connect to GCP Pub/Sub APIs.
  *
  * @author Vinicius Carvalho
+ * @author Patrik Hörlin
  *
  * @since 1.2.2
  */
 public class PubSubHealthIndicator extends AbstractHealthIndicator {
 
-	private final PubSubTemplate pubSubTemplate;
+	private final PubSubHealthTemplate pubSubHealthTemplate;
 
-	public PubSubHealthIndicator(PubSubTemplate pubSubTemplate) {
+	public PubSubHealthIndicator(PubSubHealthTemplate pubSubHealthTemplate) {
 		super("Failed to connect to Pub/Sub APIs. Check your credentials and verify you have proper access to the service.");
-		Assert.notNull(pubSubTemplate, "PubSubTemplate can't be null");
-		this.pubSubTemplate = pubSubTemplate;
+		Assert.notNull(pubSubHealthTemplate, "PubSubHealthTemplate can't be null");
+		this.pubSubHealthTemplate = pubSubHealthTemplate;
 	}
 
 	@Override
 	protected void doHealthCheck(Health.Builder builder) {
+		ListenableFuture<List<AcknowledgeablePubsubMessage>> future = this.pubSubHealthTemplate.pullAsync();
 		try {
-			this.pubSubTemplate.pull("subscription-" + UUID.randomUUID().toString(), 1, true);
+			future.get(this.pubSubHealthTemplate.getTimeoutMillis(), TimeUnit.MILLISECONDS);
 		}
-		catch (ApiException aex) {
-			Code errorCode = aex.getStatusCode().getCode();
-			if (errorCode == StatusCode.Code.NOT_FOUND || errorCode == Code.PERMISSION_DENIED) {
-				builder.up();
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			builder.withException(e).unknown();
+		}
+		catch (ExecutionException e) {
+			Throwable t = e.getCause();
+			if (t instanceof ApiException) {
+				ApiException aex = (ApiException) t;
+				Code errorCode = aex.getStatusCode().getCode();
+				if (errorCode == StatusCode.Code.NOT_FOUND || errorCode == Code.PERMISSION_DENIED) {
```

Comment:
Hmm... Given that we're now encouraging "bring your own subscription to use for health checks", do we really want to say that a not found is ok? Seems to me like this would mask a misconfiguration error, and the error would be swallowed

Reply:
I don't know if we want to *require* users to bring their own subscription. This certainly would need to be documented well to clear up the confusion on why setting a custom subscription for health check if optional.

That being said, if the user provides a subscription, `NOT_FOUND` or `PERMISSION_DENIED` would indeed be error conditions.

Reply:
I agree that custom subscription should be optional -- for backwards compatibility, and for users who can't create a dedicated subscription due to permission structure.

So there are two paths through the healthcheck, which I'd probably handle as follows:
1) a random subscription expected to return NOT_FOUND or PERMISSION_DENIED. Any other exception would signal downtime.
2) a custom subscription, for which the expected "up" signal is *not getting any exception*. Either receiving messages or pulling an empty batch would be "up". But getting not found OR permission denied should be unexpected. My reasoning is that if a team/org has enough permissions to create a dedicated subscription, they are likely able to set permissions appropriately.

Reply:
The if/else will become even more unwieldy with my suggestions, btw, so you may want a helper method.